### PR TITLE
Add support for GPIO internal pull up/down configuration on RPi4 family.

### PR DESCRIPTION
**Pull request by tetsuya-uemura**
BCM2711 SoC on 4th generation Raspberry Pi changed the way to configure its GPIO pins' internal pull up/down resistors. NetBSD already have support for this change, and port it to FreeBSD is trivial.

This patch, based on the NetBSD commit adds the appropriate method for BCM2711 and now we can properly configure the GPIO pins' pull status.

PR: 256372
Reference: https://github.com/NetBSD/src/commit/bb88cfa64ad8045456ec4afef04a834d067c8cb1

### CODE REVIEW

File: sys/arm/broadcom/bcm2835/bcm2835_gpio.c [hunk 1]
```diff
@@ -289,16 +301,41 @@ bcm_gpio_set_function(struct bcm_gpio_softc *sc, uint32_t pin, uint32_t f)
...
-
 	/* Must be called with lock held. */
 	BCM_GPIO_LOCK_ASSERT(sc);
 
-	bank = BCM_GPIO_BANK(pin);
-	BCM_GPIO_WRITE(sc, BCM_GPIO_GPPUD(0), state);
-	BCM_GPIO_WRITE(sc, BCM_GPIO_GPPUDCLK(bank), BCM_GPIO_MASK(pin));
-	BCM_GPIO_WRITE(sc, BCM_GPIO_GPPUD(0), 0);
-	BCM_GPIO_WRITE(sc, BCM_GPIO_GPPUDCLK(bank), 0);
+	if (sc->sc_is2711 == false) { /* BCM2835 */
```
Comment (gmshake) [inline sys/arm/broadcom/bcm2835/bcm2835_gpio.c:307]: 'sc_is2711' is boolean so no need to compare it to false.

if (!sc->sc_is2711) {} is more readable.

Comment (emaste) [reply sys/arm/broadcom/bcm2835/bcm2835_gpio.c:307]: Or put the true case first
```
if (sc->sc_is2711) {
		u_int mask  = BCM2711_GPIO_MASK(pin);
		u_int regid = BCM2711_GPIO_REGID(pin);
		uint32_t reg;
                ...
} else { /* BCM2835 */
...
}
```

File: sys/arm/broadcom/bcm2835/bcm2835_gpio.c [hunk 2]
```diff
@@ -770,24 +809,28 @@ bcm_gpio_attach(device_t dev)
...
-		device_printf(dev, "unable to setup the gpio irq handler\n");
-		goto fail;
-	}
 	/* Find our node. */
 	gpio = ofw_bus_get_node(sc->sc_dev);
 	if (!OF_hasprop(gpio, "gpio-controller"))
 		/* Node is not a GPIO controller. */
 		goto fail;
+	/* Guess I'm BCM2711 or not. */
+	sc->sc_is2711 = ofw_bus_node_is_compatible(gpio, "brcm,bcm2711-gpio")
```
Comment (emaste) [inline sys/arm/broadcom/bcm2835/bcm2835_gpio.c:818]: Dropping the `? true : false` is clearer IMO
`sc->sc_is2711 = ofw_bus_node_is_compatible(gpio, "brcm,bcm2711-gpio");`

File: sys/arm/broadcom/bcm2835/bcm2835_gpio.c [hunk 3]
```diff
@@ -151,8 +159,12 @@ enum bcm_gpio_pud {
 #define	BCM_GPIO_GPLEN(_bank)	(0x70 + _bank * 4)	/* Low Level irq */
 #define	BCM_GPIO_GPAREN(_bank)	(0x7c + _bank * 4)	/* Async Rising Edge */
 #define	BCM_GPIO_GPAFEN(_bank)	(0x88 + _bank * 4)	/* Async Falling Egde */
-#define	BCM_GPIO_GPPUD(_bank)	(0x94)			/* Pin Pull up/down */
-#define	BCM_GPIO_GPPUDCLK(_bank) (0x98 + _bank * 4)	/* Pin Pull up clock */
+#define	BCM2835_GPIO_GPPUD(_bank) (0x94)		/* Pin Pull up/down */
+#define	BCM2835_GPIO_GPPUDCLK(_bank) (0x98 + _bank * 4)	/* Pin Pull up clock */
+
+#define	BCM2711_GPIO_GPPUD(x)	(0x0e4 + (x) * sizeof(uint32_t)) /* Pin Pull up/down */
+#define	BCM2711_GPIO_MASK(n)	(0x3 << ((n) % 16)*2)
```
Comment (mhorne) [inline sys/arm/broadcom/bcm2835/bcm2835_gpio.c:166]: If you add:
```
#define BCM2711_GPIO_PIN_SHIFT(n) (((n) % 16) * 2)
```

Then you can use it as part of this definition, and below (see my other comment).

**Review by mhorne — COMMENTED** [review]
Looking good, almost there.

File: sys/arm/broadcom/bcm2835/bcm2835_gpio.c [hunk 4]
```diff
@@ -109,6 +115,8 @@ struct bcm_gpio_softc {
 	device_t		sc_busdev;
 	struct mtx		sc_mtx;
 	struct resource *	sc_res[BCM_GPIO_IRQS + 1];
+	bool			sc_is2711;
+	u_int			sc_maxpins;
```
Comment (mhorne) [inline sys/arm/broadcom/bcm2835/bcm2835_gpio.c:119]: `sc_res` and `sc_bst`/`sc_bsh` are functionally related, so we shouldn't insert the new struct members between them.

I would suggest placing the new ones right above `sc_gpio_npins`.

File: sys/arm/broadcom/bcm2835/bcm2835_gpio.c [hunk 5]
```diff
@@ -289,16 +302,39 @@ bcm_gpio_set_function(struct bcm_gpio_softc *sc, uint32_t pin, uint32_t f)
...
+		    case BCM2835_PUD_OFF:
+			state = BCM2711_PUD_OFF;
+			break;
+		    case BCM2835_PUD_DOWN:
+			state = BCM2711_PUD_DOWN;
+			break;
+		    case BCM2835_PUD_UP:
+			state = BCM2711_PUD_UP;
+			break;
+		}
```
Comment (mhorne) [inline sys/arm/broadcom/bcm2835/bcm2835_gpio.c:323]: We have an unusual style for `switch` statements in FreeBSD, in which the `case` lines should not be indented. Meaning:
```
switch (state) {
case BCM2835_PUD_OFF:
        state = BCM2711_PUD_OFF;
        break;
case BCM2835_PUD_DOWN:
...
```

**Review by mhorne — COMMENTED** [review]
Sorry, two more small things to fix that I should have noticed before. After that, I will be able to merge this.

**Review by mhorne — APPROVED** [review]

**Comment by mhorne** [comment]
I have merged the change, so this PR can be closed. It does not seem like I am able to do it myself.

**Comment by tetsuya-uemura** [comment]
Thanks! Changes had been merged and this request should be closed.

**Pull request closed without merging**